{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ae919fc8-a0b3-490a-84a7-583aa8030dcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pickle\n",
    "\n",
    "import pyensembl\n",
    "\n",
    "pd.set_option('display.max_columns', None)\n",
    "os.chdir(\"/home/ylee/blast/blastdb/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "292696e5-288c-4975-a802-6d056e1de3f0",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "genome = pyensembl.EnsemblRelease(100, \"mus_musculus\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "add5541f-c1a5-4327-a49a-98a88f7a909f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load BLAST results into a pandas DataFrame\n",
    "blast_Actg1 = pd.read_csv(\"blast_Actg1_local_wss7.table\", sep=\"\\t\", comment='#', header=None)\n",
    "# blast_Actg1 = pd.read_csv(\"blast_Actg1_local_ws4_eval5_1.table\", sep=\"\\t\", comment='#', header=None)\n",
    "# Define column names\n",
    "# [\"query id\", \"subject id\", \"% identity\", \"alignment length\", \"q. start\", \"q. end\", \"query length\", \"s. start\", \"s. end\", \"subject strand\", \"subject length\", \"mismatches\", \"gaps\", \"evalue\", \"bit score\"]\n",
    "blast_Actg1.columns = [\"qseqid\", \"sseqid\", \"pident\", \"length\", \"qstart\", \"qend\", \"qlen\", \n",
    "                                 \"sstart\", \"send\", \"sstrand\", \"slen\", \"mismatch\", \"gaps\", \"evalue\", \"bitscore\"]\n",
    "\n",
    "blast_Actg1['genes'] = pd.NA\n",
    "blast_Actg1['extend'] = pd.NA\n",
    "\n",
    "blast_Actg1 = blast_Actg1.head(270)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "50b55125-2a23-4385-b0e5-478e97e9f93c",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>pident</th>\n",
       "      <th>length</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>qlen</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>sstrand</th>\n",
       "      <th>slen</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gaps</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>genes</th>\n",
       "      <th>extend</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUST00000071555.12</td>\n",
       "      <td>chr4</td>\n",
       "      <td>99.738</td>\n",
       "      <td>1905</td>\n",
       "      <td>48</td>\n",
       "      <td>1952</td>\n",
       "      <td>1952</td>\n",
       "      <td>103563325</td>\n",
       "      <td>103565229</td>\n",
       "      <td>plus</td>\n",
       "      <td>156508116</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3491.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUST00000071555.12</td>\n",
       "      <td>chr4</td>\n",
       "      <td>77.778</td>\n",
       "      <td>99</td>\n",
       "      <td>1120</td>\n",
       "      <td>1216</td>\n",
       "      <td>1952</td>\n",
       "      <td>56740228</td>\n",
       "      <td>56740133</td>\n",
       "      <td>minus</td>\n",
       "      <td>156508116</td>\n",
       "      <td>17</td>\n",
       "      <td>5</td>\n",
       "      <td>0.000051</td>\n",
       "      <td>56.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUST00000071555.12</td>\n",
       "      <td>chr4</td>\n",
       "      <td>94.444</td>\n",
       "      <td>36</td>\n",
       "      <td>754</td>\n",
       "      <td>789</td>\n",
       "      <td>1952</td>\n",
       "      <td>154667036</td>\n",
       "      <td>154667001</td>\n",
       "      <td>minus</td>\n",
       "      <td>156508116</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000051</td>\n",
       "      <td>56.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUST00000071555.12</td>\n",
       "      <td>chr4</td>\n",
       "      <td>72.671</td>\n",
       "      <td>161</td>\n",
       "      <td>350</td>\n",
       "      <td>505</td>\n",
       "      <td>1952</td>\n",
       "      <td>154667440</td>\n",
       "      <td>154667285</td>\n",
       "      <td>minus</td>\n",
       "      <td>156508116</td>\n",
       "      <td>34</td>\n",
       "      <td>10</td>\n",
       "      <td>0.110000</td>\n",
       "      <td>45.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUST00000071555.12</td>\n",
       "      <td>chr4</td>\n",
       "      <td>78.125</td>\n",
       "      <td>64</td>\n",
       "      <td>582</td>\n",
       "      <td>645</td>\n",
       "      <td>1952</td>\n",
       "      <td>154667208</td>\n",
       "      <td>154667145</td>\n",
       "      <td>minus</td>\n",
       "      <td>156508116</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>1.400000</td>\n",
       "      <td>41.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>ENSMUST00000071555.12</td>\n",
       "      <td>chr3</td>\n",
       "      <td>72.680</td>\n",
       "      <td>194</td>\n",
       "      <td>1703</td>\n",
       "      <td>1893</td>\n",
       "      <td>1952</td>\n",
       "      <td>83390771</td>\n",
       "      <td>83390583</td>\n",
       "      <td>minus</td>\n",
       "      <td>160039680</td>\n",
       "      <td>45</td>\n",
       "      <td>8</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>58.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>ENSMUST00000071555.12</td>\n",
       "      <td>chr3</td>\n",
       "      <td>84.615</td>\n",
       "      <td>52</td>\n",
       "      <td>1902</td>\n",
       "      <td>1952</td>\n",
       "      <td>1952</td>\n",
       "      <td>83390072</td>\n",
       "      <td>83390022</td>\n",
       "      <td>minus</td>\n",
       "      <td>160039680</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>0.002000</td>\n",
       "      <td>51.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267</th>\n",
       "      <td>ENSMUST00000071555.12</td>\n",
       "      <td>chr3</td>\n",
       "      <td>78.689</td>\n",
       "      <td>61</td>\n",
       "      <td>1136</td>\n",
       "      <td>1196</td>\n",
       "      <td>1952</td>\n",
       "      <td>30597950</td>\n",
       "      <td>30597890</td>\n",
       "      <td>minus</td>\n",
       "      <td>160039680</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>1.400000</td>\n",
       "      <td>41.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>268</th>\n",
       "      <td>ENSMUST00000071555.12</td>\n",
       "      <td>chr3</td>\n",
       "      <td>85.000</td>\n",
       "      <td>40</td>\n",
       "      <td>1805</td>\n",
       "      <td>1843</td>\n",
       "      <td>1952</td>\n",
       "      <td>123326807</td>\n",
       "      <td>123326845</td>\n",
       "      <td>plus</td>\n",
       "      <td>160039680</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>5.100000</td>\n",
       "      <td>39.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>269</th>\n",
       "      <td>ENSMUST00000071555.12</td>\n",
       "      <td>chrY</td>\n",
       "      <td>95.833</td>\n",
       "      <td>24</td>\n",
       "      <td>1172</td>\n",
       "      <td>1195</td>\n",
       "      <td>1952</td>\n",
       "      <td>5417418</td>\n",
       "      <td>5417395</td>\n",
       "      <td>minus</td>\n",
       "      <td>91744698</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>5.100000</td>\n",
       "      <td>39.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>270 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    qseqid sseqid  pident  length  qstart  qend  qlen  \\\n",
       "0    ENSMUST00000071555.12   chr4  99.738    1905      48  1952  1952   \n",
       "1    ENSMUST00000071555.12   chr4  77.778      99    1120  1216  1952   \n",
       "2    ENSMUST00000071555.12   chr4  94.444      36     754   789  1952   \n",
       "3    ENSMUST00000071555.12   chr4  72.671     161     350   505  1952   \n",
       "4    ENSMUST00000071555.12   chr4  78.125      64     582   645  1952   \n",
       "..                     ...    ...     ...     ...     ...   ...   ...   \n",
       "265  ENSMUST00000071555.12   chr3  72.680     194    1703  1893  1952   \n",
       "266  ENSMUST00000071555.12   chr3  84.615      52    1902  1952  1952   \n",
       "267  ENSMUST00000071555.12   chr3  78.689      61    1136  1196  1952   \n",
       "268  ENSMUST00000071555.12   chr3  85.000      40    1805  1843  1952   \n",
       "269  ENSMUST00000071555.12   chrY  95.833      24    1172  1195  1952   \n",
       "\n",
       "        sstart       send sstrand       slen  mismatch  gaps    evalue  \\\n",
       "0    103563325  103565229    plus  156508116         5     0  0.000000   \n",
       "1     56740228   56740133   minus  156508116        17     5  0.000051   \n",
       "2    154667036  154667001   minus  156508116         2     0  0.000051   \n",
       "3    154667440  154667285   minus  156508116        34    10  0.110000   \n",
       "4    154667208  154667145   minus  156508116        14     0  1.400000   \n",
       "..         ...        ...     ...        ...       ...   ...       ...   \n",
       "265   83390771   83390583   minus  160039680        45     8  0.000014   \n",
       "266   83390072   83390022   minus  160039680         6     2  0.002000   \n",
       "267   30597950   30597890   minus  160039680        13     0  1.400000   \n",
       "268  123326807  123326845    plus  160039680         4     2  5.100000   \n",
       "269    5417418    5417395   minus   91744698         1     0  5.100000   \n",
       "\n",
       "     bitscore genes extend  \n",
       "0      3491.0   NaN    NaN  \n",
       "1        56.5   NaN    NaN  \n",
       "2        56.5   NaN    NaN  \n",
       "3        45.4   NaN    NaN  \n",
       "4        41.7   NaN    NaN  \n",
       "..        ...   ...    ...  \n",
       "265      58.4   NaN    NaN  \n",
       "266      51.0   NaN    NaN  \n",
       "267      41.7   NaN    NaN  \n",
       "268      39.9   NaN    NaN  \n",
       "269      39.9   NaN    NaN  \n",
       "\n",
       "[270 rows x 17 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blast_Actg1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "853a2d52-a43e-4767-a405-7fef69b7852c",
   "metadata": {},
   "source": [
    "## Function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2e6fb492-7628-482f-b594-952ffcb3fb7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# function that return gene IDs in locus\n",
    "def find_gene_in_locus(chrNum, start, end, strand):\n",
    "    gene_ids_at_locus = genome.gene_ids_at_locus(contig=chrNum, position=start, end=end)   # genes_at_locus is also possible\n",
    "    gene_ids_at_locus += genome.gene_ids_at_locus(contig=chrNum, position=start)\n",
    "    gene_ids_at_locus += genome.gene_ids_at_locus(contig=chrNum, position=end)\n",
    "    \n",
    "    return gene_ids_at_locus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7b3a7c4f-6e0a-4d61-821e-c1ff899b71c2",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "chromosome_sizes = {\n",
    "    \"1\": 195471971,\n",
    "    \"2\": 182113224,\n",
    "    \"3\": 160039680,\n",
    "    \"4\": 156508116,\n",
    "    \"5\": 151834684,\n",
    "    \"6\": 149736546,\n",
    "    \"7\": 145441459,\n",
    "    \"8\": 129401213,\n",
    "    \"9\": 124595110,\n",
    "    \"10\": 130694993,\n",
    "    \"11\": 122082543,\n",
    "    \"12\": 120129022,\n",
    "    \"13\": 120421639,\n",
    "    \"14\": 124902244,\n",
    "    \"15\": 104043685,\n",
    "    \"16\": 98207768,\n",
    "    \"17\": 94987271,\n",
    "    \"18\": 90702639,\n",
    "    \"19\": 61431566,\n",
    "    \"X\": 171031299,\n",
    "    \"Y\": 91744698,\n",
    "    \"MT\": 16299\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3fa63f67-f72a-4fe1-a881-6d4c8e708f7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# function that extends hit locus\n",
    "def extend_hit(chrNum, start, end, strand, extendLength = 1000):\n",
    "    chr_length = chromosome_sizes[chrNum]                              \n",
    "    move_start = start - extendLength\n",
    "    move_end = end + extendLength\n",
    "    annotation = \"both\"\n",
    "\n",
    "    if move_start < 1 and move_end > chr_length:\n",
    "        return [1, chr_length, \"last extension\"]\n",
    "    else:\n",
    "        if move_start < 1: \n",
    "            move_start = 1\n",
    "            annotation = \"upstream\"\n",
    "        elif move_end > chr_length: \n",
    "            move_end = chr_length\n",
    "            annotation = \"downstream\"\n",
    "            \n",
    "    return [move_start, move_end, annotation]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ee4ee747-9793-465e-87f4-2d1cadecb175",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene의 TSS, and distance\n",
    "def check_distance(gene_id, hit_locus):\n",
    "    \n",
    "    # create locus from gene_id\n",
    "    gene = genome.gene_by_id(gene_id)\n",
    "    gene_locus = pyensembl.Locus(contig=gene.contig, start=gene.start, end=gene.end, strand=gene.strand)\n",
    "    if gene.strand == \"+\":\n",
    "        TSS_point = gene.start\n",
    "    else:    \n",
    "        TSS_point = gene.end\n",
    "\n",
    "    # calculate the distance\n",
    "    # distance = hit_locus.distance_to_locus(gene_locus)              # if strand is different, (+/-) they calculate distance as infinity.\n",
    "    if hit_locus.distance_to_interval(gene.start, gene.end) == 0:\n",
    "        distance = 0\n",
    "    else:\n",
    "        distance = hit_locus.distance_to_interval(TSS_point, TSS_point)\n",
    "    \n",
    "    return distance"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "591c546c-64b9-4505-9882-8fc923d2ef64",
   "metadata": {},
   "source": [
    "## Main - add 2kb distance at hit locus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "11cde61a-2779-4a5d-9cb2-415f83e5e7b9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48\n"
     ]
    }
   ],
   "source": [
    "similar_pcg_set = set()\n",
    "sigupregul_pcg_rows = []\n",
    "appended_indices = {\n",
    "    \"ENSMUSG00000020246\": set(),\n",
    "    \"ENSMUSG00000023868\": set(),\n",
    "    \"ENSMUSG00000035783\": set(),\n",
    "    \"ENSMUSG00000037351\": set(),\n",
    "    \"ENSMUSG00000047497\": set(),\n",
    "    \"ENSMUSG00000059430\": set(),\n",
    "    \"ENSMUSG00000059895\": set()\n",
    "}\n",
    "\n",
    "i = 0\n",
    "while i < len(blast_Actg1): \n",
    "    row = blast_Actg1.iloc[i]\n",
    "\n",
    "    hit_chrNum = row['sseqid'][3:]\n",
    "    if row['sstrand']=='plus':\n",
    "        hit_start = int(row['sstart'])\n",
    "        hit_end =  int(row['send'])\n",
    "        hit_strand = \"+\" \n",
    "    else:\n",
    "        hit_start = int(row['send'])\n",
    "        hit_end =  int(row['sstart'])\n",
    "        hit_strand = \"-\"\n",
    "\n",
    "    # create hit Locus object\n",
    "    hit_locus = pyensembl.Locus(contig=hit_chrNum, start=hit_start, end=hit_end, strand=hit_strand)\n",
    "\n",
    "    extend = \"NA\"\n",
    "    while (extend != \"false\"):\n",
    "        extend_hit_result = extend_hit(hit_chrNum, hit_start, hit_end, hit_strand, 2000)    # default extension\n",
    "        hit_start = extend_hit_result[0]\n",
    "        hit_end = extend_hit_result[1]\n",
    "        gene_in_locus_list = find_gene_in_locus(hit_chrNum, hit_start, hit_end, hit_strand)             # find gene in locus\n",
    "        if len(gene_in_locus_list) == 0:                                                                # if there is no gene in locus, extend hit length\n",
    "            extend_hit_result = extend_hit(hit_chrNum, hit_start, hit_end, hit_strand, 1000)\n",
    "            hit_start = extend_hit_result[0]\n",
    "            hit_end = extend_hit_result[1]\n",
    "            extend = extend_hit_result[2]\n",
    "        else:                                                                                            # if there is gene in locus, print the gene id and distance\n",
    "            # check distance between hit locus and gene\n",
    "            info = \"\"\n",
    "            for gene_id in gene_in_locus_list: \n",
    "                distance = check_distance(gene_id, hit_locus)\n",
    "                gene = genome.gene_by_id(gene_id);\n",
    "                info += f\"{gene_id} - Distance: {distance}; \"\n",
    "                ####################################\n",
    "                # print(gene_id, distance)    \n",
    "                # if (gene_id == 'ENSMUSG00000070980'):\n",
    "                #    print(\"Check ENSMUSG00000096930 distance : \", check_distance(\"ENSMUSG00000096930\", hit_locus), \"\\n\",\n",
    "                #          \"Hit start/end : \", hit_locus.start, hit_locus.end, \"\\n\",\n",
    "                #          \"Extended Hit start/end : \", hit_start, hit_end, \"\\n\",\n",
    "                #          \"Gene strand/start/end : \", genome.gene_by_id('ENSMUSG00000096930').strand, genome.gene_by_id('ENSMUSG00000096930').start, genome.gene_by_id('ENSMUSG00000096930').end)\n",
    "                ####################################\n",
    "                if distance < 2000 and gene.biotype == \"protein_coding\":\n",
    "                    similar_pcg_set.add(gene_id)\n",
    "                    \n",
    "                    #################################\n",
    "                    if gene.gene_id == \"ENSMUSG00000020246\":  # Hcfc2\n",
    "                        if i not in appended_indices[\"ENSMUSG00000020246\"]:\n",
    "                            row_copy = row.copy()\n",
    "                            row_copy['gene_id'] = \"ENSMUSG00000020246\"\n",
    "                            row_copy['gene_name'] = \"Hcfc2\"\n",
    "                            sigupregul_pcg_rows.append(row_copy)\n",
    "                            appended_indices[\"ENSMUSG00000020246\"].add(i)\n",
    "                    \n",
    "                    if gene.gene_id == \"ENSMUSG00000023868\":  # Pde10a\n",
    "                        if i not in appended_indices[\"ENSMUSG00000023868\"]:\n",
    "                            row_copy = row.copy()\n",
    "                            row_copy['gene_id'] = \"ENSMUSG00000023868\"\n",
    "                            row_copy['gene_name'] = \"Pde10a\"\n",
    "                            sigupregul_pcg_rows.append(row_copy)\n",
    "                            appended_indices[\"ENSMUSG00000023868\"].add(i)\n",
    "\n",
    "                    if gene.gene_id == \"ENSMUSG00000035783\":  # Acta2\n",
    "                        if i not in appended_indices[\"ENSMUSG00000035783\"]:\n",
    "                            row_copy = row.copy()\n",
    "                            row_copy['gene_id'] = \"ENSMUSG00000035783\"\n",
    "                            row_copy['gene_name'] = \"Acta2\"\n",
    "                            sigupregul_pcg_rows.append(row_copy)\n",
    "                            appended_indices[\"ENSMUSG00000035783\"].add(i)\n",
    "                    \n",
    "                    if gene.gene_id == \"ENSMUSG00000037351\":  # Actr1b\n",
    "                        if i not in appended_indices[\"ENSMUSG00000037351\"]:\n",
    "                            row_copy = row.copy()\n",
    "                            row_copy['gene_id'] = \"ENSMUSG00000037351\"\n",
    "                            row_copy['gene_name'] = \"Actr1b\"\n",
    "                            sigupregul_pcg_rows.append(row_copy)\n",
    "                            appended_indices[\"ENSMUSG00000037351\"].add(i)\n",
    "                    \n",
    "                    if gene.gene_id == \"ENSMUSG00000047497\":  # Adamts12\n",
    "                        if i not in appended_indices[\"ENSMUSG00000047497\"]:\n",
    "                            row_copy = row.copy()\n",
    "                            row_copy['gene_id'] = \"ENSMUSG00000047497\"\n",
    "                            row_copy['gene_name'] = \"Adamts12\"\n",
    "                            sigupregul_pcg_rows.append(row_copy)\n",
    "                            appended_indices[\"ENSMUSG00000047497\"].add(i)\n",
    "\n",
    "                    if gene.gene_id == \"ENSMUSG00000059430\":  # Actg2\n",
    "                        if i not in appended_indices[\"ENSMUSG00000059430\"]:\n",
    "                            row_copy = row.copy()\n",
    "                            row_copy['gene_id'] = \"ENSMUSG00000059430\"\n",
    "                            row_copy['gene_name'] = \"Actg2\"\n",
    "                            sigupregul_pcg_rows.append(row_copy)\n",
    "                            appended_indices[\"ENSMUSG00000059430\"].add(i)\n",
    "\n",
    "                    if gene.gene_id == \"ENSMUSG00000059895\":  # Ptp4a3\n",
    "                        if i not in appended_indices[\"ENSMUSG00000059895\"]:\n",
    "                            row_copy = row.copy()\n",
    "                            row_copy['gene_id'] = \"ENSMUSG00000059895\"\n",
    "                            row_copy['gene_name'] = \"Ptp4a3\"\n",
    "                            sigupregul_pcg_rows.append(row_copy)\n",
    "                            appended_indices[\"ENSMUSG00000059895\"].add(i)\n",
    "                    #################################\n",
    "            \n",
    "            # Update DataFrame, add gene and extend information\n",
    "            # blast_Actg1.at[i, 'genes'] = info\n",
    "            # blast_Actg1.at[i, 'extend'] = extend\n",
    "            extend = 'false'\n",
    "\n",
    "    i += 1\n",
    "\n",
    "print(len(similar_pcg_set))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "51e7db6d-9bae-4012-8c43-54efb25281b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24\n"
     ]
    }
   ],
   "source": [
    "print(len(sigupregul_pcg_rows))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "79bbe849-5fd9-4588-b585-f75f121a0816",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sigupregul_pcg = pd.DataFrame(sigupregul_pcg_rows)\n",
    "df_sigupregul_pcg.to_csv('sigupregul_pcg_rows.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "6ab623d1-6687-40b8-aabc-062cb7119f1e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "40\n"
     ]
    }
   ],
   "source": [
    "os.chdir(\"/home/ylee/blast/result/\")\n",
    "with open('similar_pcg_set_Y.pkl', 'rb') as file:\n",
    "    similar_pcg_set_Y = pickle.load(file)\n",
    "\n",
    "with open('similar_pcg_set_C.pkl', 'rb') as file:\n",
    "    similar_pcg_set_C = pickle.load(file)\n",
    "\n",
    "similar_pcg_set_Common = similar_pcg_set_C - (similar_pcg_set_C - similar_pcg_set)\n",
    "print(len(similar_pcg_set_Common))\n",
    "\n",
    "with open('similar_pcg_set_Common.pkl', 'wb') as file:\n",
    "    pickle.dump(similar_pcg_set_Common, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8deafe81-aaf1-4bc6-a6d5-14e489ab1205",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"/home/ylee/blast/result/\")\n",
    "\n",
    "pcg_ids = [gene.gene_id for gene in genome.genes() if gene.biotype == 'protein_coding']\n",
    "pcg_set = set(pcg_ids)\n",
    "nonsimilar_pcg_set = pcg_set - similar_pcg_set\n",
    "\n",
    "# Save the set as a pickle file\n",
    "with open('similar_pcg_set_Y.pkl', 'wb') as file:\n",
    "    pickle.dump(similar_pcg_set, file)\n",
    "\n",
    "with open('nonsimilar_pcg_set_Y.pkl', 'wb') as file:\n",
    "    pickle.dump(nonsimilar_pcg_set, file)\n",
    "\n",
    "with open('pcg_set.pkl', 'wb') as file:\n",
    "    pickle.dump(pcg_set, file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "219607cb-309f-47c6-b462-94c0a59317d5",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Codes from K"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "06a8bbb0-218c-4ef0-b57d-baaeaaf36557",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_overlapped(start_1, end_1, start_2, end_2):\n",
    "    return max(start_1, start_2) <= min(end_1, end_2)\n",
    "\n",
    "def check_TSS_below_2kb(gene, start, end):\n",
    "    if gene.strand == '-':\n",
    "        TSS = gene.end\n",
    "    else:\n",
    "        TSS = gene.start\n",
    "        \n",
    "    if start <= TSS <= end:\n",
    "        distance =  0\n",
    "    elif TSS < start:\n",
    "        distance = start - TSS\n",
    "    else:\n",
    "        distance = TSS - end\n",
    "\n",
    "    return distance <= 2000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "bebe0c6b-b8a2-4b50-b8a1-37cda50916ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48\n"
     ]
    }
   ],
   "source": [
    "similar_pcg_set_K = set()\n",
    "\n",
    "i = 0\n",
    "while i < len(blast_Actg1_local_ws4): \n",
    "    row = blast_Actg1_local_ws4.iloc[i]\n",
    "\n",
    "    hit_chrNum = row['sseqid'][3:]\n",
    "    if row['sstrand']=='plus':\n",
    "        hit_start = int(row['sstart'])\n",
    "        hit_end =  int(row['send'])\n",
    "        hit_strand = \"+\" \n",
    "    else:\n",
    "        hit_start = int(row['send'])\n",
    "        hit_end =  int(row['sstart'])\n",
    "        hit_strand = \"-\"\n",
    "\n",
    "    # create hit Locus object\n",
    "    hit_locus = pyensembl.Locus(contig=hit_chrNum, start=hit_start, end=hit_end, strand=hit_strand)\n",
    "\n",
    "    extend = \"NA\"\n",
    "    while (extend != \"false\"):\n",
    "        extend_hit_result = extend_hit(hit_chrNum, hit_start, hit_end, hit_strand, 2000)    # default extension\n",
    "        hit_start = extend_hit_result[0]\n",
    "        hit_end = extend_hit_result[1]\n",
    "        gene_in_locus_list = find_gene_in_locus(hit_chrNum, hit_start, hit_end, hit_strand)             # find gene in locus\n",
    "        if len(gene_in_locus_list) == 0:                                                                # if there is no gene in locus, extend hit length\n",
    "            extend_hit_result = extend_hit(hit_chrNum, hit_start, hit_end, hit_strand, 1000)\n",
    "            hit_start = extend_hit_result[0]\n",
    "            hit_end = extend_hit_result[1]\n",
    "            extend = extend_hit_result[2]\n",
    "        else:\n",
    "            for gene_id in gene_in_locus_list: \n",
    "                gene = genome.gene_by_id(gene_id);\n",
    "                check1 = check_overlapped(gene.start, gene.end, hit_locus.start, hit_locus.end);\n",
    "                check2 = check_TSS_below_2kb(gene, hit_locus.start, hit_locus.end);\n",
    "                if check1 or check2:\n",
    "                    if gene.biotype == \"protein_coding\":\n",
    "                        similar_pcg_set_K.add(gene_id)\n",
    "            extend = 'false'\n",
    "\n",
    "    i += 1\n",
    "\n",
    "print(len(similar_pcg_set_K))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py_3.8.13",
   "language": "python",
   "name": "py_3.8.13."
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
