{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "082f12e9-54fe-45da-bbe9-853bfa19b31e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pickle\n",
    "\n",
    "import pyensembl\n",
    "\n",
    "pd.set_option('display.max_columns', None)\n",
    "os.chdir(\"/home/ylee/blast/result/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3f1d0130-0a12-4d67-9ca9-6a59bde64b6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = pyensembl.EnsemblRelease(100, \"mus_musculus\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "44921381-1cfb-4ffd-9714-474b6467c9d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "blast_Actg1 = pd.read_csv(\"ENSMUST00000071555.12.transcript.blastn.tsv.parsed.anno.final\", sep=\"\\t\", header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "911de082-eef6-405d-bc5c-c8e28ed82071",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query id</th>\n",
       "      <th>query gene_id</th>\n",
       "      <th>query gene_symbol</th>\n",
       "      <th>subject id</th>\n",
       "      <th>% identity</th>\n",
       "      <th>alignment length</th>\n",
       "      <th>q. start</th>\n",
       "      <th>q. end</th>\n",
       "      <th>q. strand</th>\n",
       "      <th>query length</th>\n",
       "      <th>s. start</th>\n",
       "      <th>s. end</th>\n",
       "      <th>subject strand</th>\n",
       "      <th>subject length</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>gaps</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bit score</th>\n",
       "      <th>feature</th>\n",
       "      <th>feat_start</th>\n",
       "      <th>feat_end</th>\n",
       "      <th>feat_strand</th>\n",
       "      <th>feat_ovl_peak</th>\n",
       "      <th>peak_ovl_feat</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUST00000071555.12</td>\n",
       "      <td>ENSMUSG00000062825</td>\n",
       "      <td>Actg1</td>\n",
       "      <td>chr10</td>\n",
       "      <td>88.93</td>\n",
       "      <td>1039</td>\n",
       "      <td>51</td>\n",
       "      <td>1085</td>\n",
       "      <td>-</td>\n",
       "      <td>1946</td>\n",
       "      <td>25053288</td>\n",
       "      <td>25054301</td>\n",
       "      <td>+</td>\n",
       "      <td>130694993</td>\n",
       "      <td>86</td>\n",
       "      <td>29</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1254.0</td>\n",
       "      <td>gene,gene,promoter_proximal,promoter_proximal</td>\n",
       "      <td>25,053,359,250,541,400,000,000,000,000,000</td>\n",
       "      <td>25,054,685,250,542,200,000,000,000,000,000</td>\n",
       "      <td>+,+,+,+</td>\n",
       "      <td>0.93,0.09,0.85,0.08</td>\n",
       "      <td>0.72,1.0,0.43,0.04</td>\n",
       "      <td>ENSMUSG00000112597,ENSMUSG00000093282,ENSMUSG0...</td>\n",
       "      <td>AC153974.2,Gm22566,Gm22566,AC153974.2</td>\n",
       "      <td>processed_pseudogene,miRNA,miRNA,processed_pse...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUST00000071555.12</td>\n",
       "      <td>ENSMUSG00000062825</td>\n",
       "      <td>Actg1</td>\n",
       "      <td>chr10</td>\n",
       "      <td>91.09</td>\n",
       "      <td>875</td>\n",
       "      <td>1081</td>\n",
       "      <td>1945</td>\n",
       "      <td>-</td>\n",
       "      <td>1946</td>\n",
       "      <td>33893385</td>\n",
       "      <td>33894256</td>\n",
       "      <td>-</td>\n",
       "      <td>130694993</td>\n",
       "      <td>65</td>\n",
       "      <td>13</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1171.0</td>\n",
       "      <td>gene</td>\n",
       "      <td>33894093</td>\n",
       "      <td>33894256</td>\n",
       "      <td>-</td>\n",
       "      <td>0.19</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000112534</td>\n",
       "      <td>AC153961.2</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUST00000071555.12</td>\n",
       "      <td>ENSMUSG00000062825</td>\n",
       "      <td>Actg1</td>\n",
       "      <td>chr10</td>\n",
       "      <td>92.13</td>\n",
       "      <td>635</td>\n",
       "      <td>1316</td>\n",
       "      <td>1946</td>\n",
       "      <td>-</td>\n",
       "      <td>1946</td>\n",
       "      <td>25054678</td>\n",
       "      <td>25055312</td>\n",
       "      <td>+</td>\n",
       "      <td>130694993</td>\n",
       "      <td>46</td>\n",
       "      <td>4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>893.0</td>\n",
       "      <td>gene</td>\n",
       "      <td>25053359</td>\n",
       "      <td>25054685</td>\n",
       "      <td>+</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.01</td>\n",
       "      <td>ENSMUSG00000112597</td>\n",
       "      <td>AC153974.2</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUST00000071555.12</td>\n",
       "      <td>ENSMUSG00000062825</td>\n",
       "      <td>Actg1</td>\n",
       "      <td>chr11</td>\n",
       "      <td>82.16</td>\n",
       "      <td>953</td>\n",
       "      <td>1020</td>\n",
       "      <td>1946</td>\n",
       "      <td>-</td>\n",
       "      <td>1946</td>\n",
       "      <td>107191427</td>\n",
       "      <td>107192345</td>\n",
       "      <td>+</td>\n",
       "      <td>122082543</td>\n",
       "      <td>110</td>\n",
       "      <td>60</td>\n",
       "      <td>0.0</td>\n",
       "      <td>763.0</td>\n",
       "      <td>gene</td>\n",
       "      <td>107191094</td>\n",
       "      <td>107191630</td>\n",
       "      <td>+</td>\n",
       "      <td>0.23</td>\n",
       "      <td>0.38</td>\n",
       "      <td>ENSMUSG00000081360</td>\n",
       "      <td>Gm11718</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUST00000071555.12</td>\n",
       "      <td>ENSMUSG00000062825</td>\n",
       "      <td>Actg1</td>\n",
       "      <td>chr11</td>\n",
       "      <td>89.20</td>\n",
       "      <td>1871</td>\n",
       "      <td>108</td>\n",
       "      <td>1946</td>\n",
       "      <td>-</td>\n",
       "      <td>1946</td>\n",
       "      <td>12936134</td>\n",
       "      <td>12937970</td>\n",
       "      <td>-</td>\n",
       "      <td>122082543</td>\n",
       "      <td>136</td>\n",
       "      <td>66</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2276.0</td>\n",
       "      <td>gene,promoter_proximal</td>\n",
       "      <td>1,293,683,412,937,930</td>\n",
       "      <td>1,293,793,812,939,930</td>\n",
       "      <td>-,-</td>\n",
       "      <td>0.61,0.02</td>\n",
       "      <td>1.0,0.02</td>\n",
       "      <td>ENSMUSG00000083859,ENSMUSG00000083859</td>\n",
       "      <td>Gm12003,Gm12003</td>\n",
       "      <td>processed_pseudogene,processed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256</th>\n",
       "      <td>ENSMUST00000071555.12</td>\n",
       "      <td>ENSMUSG00000062825</td>\n",
       "      <td>Actg1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>86.11</td>\n",
       "      <td>36</td>\n",
       "      <td>155</td>\n",
       "      <td>190</td>\n",
       "      <td>-</td>\n",
       "      <td>1946</td>\n",
       "      <td>133210507</td>\n",
       "      <td>133210542</td>\n",
       "      <td>+</td>\n",
       "      <td>171031299</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5.1</td>\n",
       "      <td>39.9</td>\n",
       "      <td>gene</td>\n",
       "      <td>133210486</td>\n",
       "      <td>133211520</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>0.04</td>\n",
       "      <td>ENSMUSG00000083858</td>\n",
       "      <td>Gm14979</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>ENSMUST00000071555.12</td>\n",
       "      <td>ENSMUSG00000062825</td>\n",
       "      <td>Actg1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>86.84</td>\n",
       "      <td>38</td>\n",
       "      <td>1820</td>\n",
       "      <td>1856</td>\n",
       "      <td>-</td>\n",
       "      <td>1946</td>\n",
       "      <td>155220562</td>\n",
       "      <td>155220596</td>\n",
       "      <td>-</td>\n",
       "      <td>171031299</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>5.1</td>\n",
       "      <td>39.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>ENSMUST00000071555.12</td>\n",
       "      <td>ENSMUSG00000062825</td>\n",
       "      <td>Actg1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>90.00</td>\n",
       "      <td>30</td>\n",
       "      <td>149</td>\n",
       "      <td>178</td>\n",
       "      <td>-</td>\n",
       "      <td>1946</td>\n",
       "      <td>133737859</td>\n",
       "      <td>133737888</td>\n",
       "      <td>-</td>\n",
       "      <td>171031299</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>5.1</td>\n",
       "      <td>39.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>ENSMUST00000071555.12</td>\n",
       "      <td>ENSMUSG00000062825</td>\n",
       "      <td>Actg1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>95.83</td>\n",
       "      <td>24</td>\n",
       "      <td>816</td>\n",
       "      <td>839</td>\n",
       "      <td>-</td>\n",
       "      <td>1946</td>\n",
       "      <td>117265226</td>\n",
       "      <td>117265249</td>\n",
       "      <td>-</td>\n",
       "      <td>171031299</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>5.1</td>\n",
       "      <td>39.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>ENSMUST00000071555.12</td>\n",
       "      <td>ENSMUSG00000062825</td>\n",
       "      <td>Actg1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>95.83</td>\n",
       "      <td>24</td>\n",
       "      <td>1167</td>\n",
       "      <td>1190</td>\n",
       "      <td>-</td>\n",
       "      <td>1946</td>\n",
       "      <td>5417395</td>\n",
       "      <td>5417418</td>\n",
       "      <td>-</td>\n",
       "      <td>91744698</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>5.1</td>\n",
       "      <td>39.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>261 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  query id       query gene_id query gene_symbol subject id  \\\n",
       "0    ENSMUST00000071555.12  ENSMUSG00000062825             Actg1      chr10   \n",
       "1    ENSMUST00000071555.12  ENSMUSG00000062825             Actg1      chr10   \n",
       "2    ENSMUST00000071555.12  ENSMUSG00000062825             Actg1      chr10   \n",
       "3    ENSMUST00000071555.12  ENSMUSG00000062825             Actg1      chr11   \n",
       "4    ENSMUST00000071555.12  ENSMUSG00000062825             Actg1      chr11   \n",
       "..                     ...                 ...               ...        ...   \n",
       "256  ENSMUST00000071555.12  ENSMUSG00000062825             Actg1       chrX   \n",
       "257  ENSMUST00000071555.12  ENSMUSG00000062825             Actg1       chrX   \n",
       "258  ENSMUST00000071555.12  ENSMUSG00000062825             Actg1       chrX   \n",
       "259  ENSMUST00000071555.12  ENSMUSG00000062825             Actg1       chrX   \n",
       "260  ENSMUST00000071555.12  ENSMUSG00000062825             Actg1       chrY   \n",
       "\n",
       "     % identity  alignment length  q. start  q. end q. strand  query length  \\\n",
       "0         88.93              1039        51    1085         -          1946   \n",
       "1         91.09               875      1081    1945         -          1946   \n",
       "2         92.13               635      1316    1946         -          1946   \n",
       "3         82.16               953      1020    1946         -          1946   \n",
       "4         89.20              1871       108    1946         -          1946   \n",
       "..          ...               ...       ...     ...       ...           ...   \n",
       "256       86.11                36       155     190         -          1946   \n",
       "257       86.84                38      1820    1856         -          1946   \n",
       "258       90.00                30       149     178         -          1946   \n",
       "259       95.83                24       816     839         -          1946   \n",
       "260       95.83                24      1167    1190         -          1946   \n",
       "\n",
       "      s. start     s. end subject strand  subject length  mismatches  gaps  \\\n",
       "0     25053288   25054301              +       130694993          86    29   \n",
       "1     33893385   33894256              -       130694993          65    13   \n",
       "2     25054678   25055312              +       130694993          46     4   \n",
       "3    107191427  107192345              +       122082543         110    60   \n",
       "4     12936134   12937970              -       122082543         136    66   \n",
       "..         ...        ...            ...             ...         ...   ...   \n",
       "256  133210507  133210542              +       171031299           5     0   \n",
       "257  155220562  155220596              -       171031299           1     4   \n",
       "258  133737859  133737888              -       171031299           3     0   \n",
       "259  117265226  117265249              -       171031299           1     0   \n",
       "260    5417395    5417418              -        91744698           1     0   \n",
       "\n",
       "     evalue  bit score                                        feature  \\\n",
       "0       0.0     1254.0  gene,gene,promoter_proximal,promoter_proximal   \n",
       "1       0.0     1171.0                                           gene   \n",
       "2       0.0      893.0                                           gene   \n",
       "3       0.0      763.0                                           gene   \n",
       "4       0.0     2276.0                         gene,promoter_proximal   \n",
       "..      ...        ...                                            ...   \n",
       "256     5.1       39.9                                           gene   \n",
       "257     5.1       39.9                                            NaN   \n",
       "258     5.1       39.9                                            NaN   \n",
       "259     5.1       39.9                                            NaN   \n",
       "260     5.1       39.9                                            NaN   \n",
       "\n",
       "                                     feat_start  \\\n",
       "0    25,053,359,250,541,400,000,000,000,000,000   \n",
       "1                                      33894093   \n",
       "2                                      25053359   \n",
       "3                                     107191094   \n",
       "4                         1,293,683,412,937,930   \n",
       "..                                          ...   \n",
       "256                                   133210486   \n",
       "257                                         NaN   \n",
       "258                                         NaN   \n",
       "259                                         NaN   \n",
       "260                                         NaN   \n",
       "\n",
       "                                       feat_end feat_strand  \\\n",
       "0    25,054,685,250,542,200,000,000,000,000,000     +,+,+,+   \n",
       "1                                      33894256           -   \n",
       "2                                      25054685           +   \n",
       "3                                     107191630           +   \n",
       "4                         1,293,793,812,939,930         -,-   \n",
       "..                                          ...         ...   \n",
       "256                                   133211520           +   \n",
       "257                                         NaN         NaN   \n",
       "258                                         NaN         NaN   \n",
       "259                                         NaN         NaN   \n",
       "260                                         NaN         NaN   \n",
       "\n",
       "           feat_ovl_peak       peak_ovl_feat  \\\n",
       "0    0.93,0.09,0.85,0.08  0.72,1.0,0.43,0.04   \n",
       "1                   0.19                   1   \n",
       "2                   0.02                0.01   \n",
       "3                   0.23                0.38   \n",
       "4              0.61,0.02            1.0,0.02   \n",
       "..                   ...                 ...   \n",
       "256                    1                0.04   \n",
       "257                  NaN                 NaN   \n",
       "258                  NaN                 NaN   \n",
       "259                  NaN                 NaN   \n",
       "260                  NaN                 NaN   \n",
       "\n",
       "                                               gene_id  \\\n",
       "0    ENSMUSG00000112597,ENSMUSG00000093282,ENSMUSG0...   \n",
       "1                                   ENSMUSG00000112534   \n",
       "2                                   ENSMUSG00000112597   \n",
       "3                                   ENSMUSG00000081360   \n",
       "4                ENSMUSG00000083859,ENSMUSG00000083859   \n",
       "..                                                 ...   \n",
       "256                                 ENSMUSG00000083858   \n",
       "257                                                NaN   \n",
       "258                                                NaN   \n",
       "259                                                NaN   \n",
       "260                                                NaN   \n",
       "\n",
       "                                 gene_name  \\\n",
       "0    AC153974.2,Gm22566,Gm22566,AC153974.2   \n",
       "1                               AC153961.2   \n",
       "2                               AC153974.2   \n",
       "3                                  Gm11718   \n",
       "4                          Gm12003,Gm12003   \n",
       "..                                     ...   \n",
       "256                                Gm14979   \n",
       "257                                    NaN   \n",
       "258                                    NaN   \n",
       "259                                    NaN   \n",
       "260                                    NaN   \n",
       "\n",
       "                                             gene_type  \n",
       "0    processed_pseudogene,miRNA,miRNA,processed_pse...  \n",
       "1                                 processed_pseudogene  \n",
       "2                                 processed_pseudogene  \n",
       "3                                 processed_pseudogene  \n",
       "4            processed_pseudogene,processed_pseudogene  \n",
       "..                                                 ...  \n",
       "256                               processed_pseudogene  \n",
       "257                                                NaN  \n",
       "258                                                NaN  \n",
       "259                                                NaN  \n",
       "260                                                NaN  \n",
       "\n",
       "[261 rows x 27 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blast_Actg1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72dc363b-d97c-4374-a1f4-74f3c2ffbeb7",
   "metadata": {},
   "source": [
    "## Function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3863c96e-e316-4596-9ceb-955410011a2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# function that return gene IDs in locus\n",
    "def find_gene_in_locus(chrNum, start, end, strand):\n",
    "    gene_ids_at_locus = genome.gene_ids_at_locus(contig=chrNum, position=start, end=end)   # genes_at_locus is also possible\n",
    "    gene_ids_at_locus += genome.gene_ids_at_locus(contig=chrNum, position=start)\n",
    "    gene_ids_at_locus += genome.gene_ids_at_locus(contig=chrNum, position=end)\n",
    "    \n",
    "    return gene_ids_at_locus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "613c5772-cdf6-4c7e-ab9c-91d09f976373",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "chromosome_sizes = {\n",
    "    \"1\": 195471971,\n",
    "    \"2\": 182113224,\n",
    "    \"3\": 160039680,\n",
    "    \"4\": 156508116,\n",
    "    \"5\": 151834684,\n",
    "    \"6\": 149736546,\n",
    "    \"7\": 145441459,\n",
    "    \"8\": 129401213,\n",
    "    \"9\": 124595110,\n",
    "    \"10\": 130694993,\n",
    "    \"11\": 122082543,\n",
    "    \"12\": 120129022,\n",
    "    \"13\": 120421639,\n",
    "    \"14\": 124902244,\n",
    "    \"15\": 104043685,\n",
    "    \"16\": 98207768,\n",
    "    \"17\": 94987271,\n",
    "    \"18\": 90702639,\n",
    "    \"19\": 61431566,\n",
    "    \"X\": 171031299,\n",
    "    \"Y\": 91744698,\n",
    "    \"MT\": 16299\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6bdc808a-a0a7-4f41-a6b3-c9b564253b2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# function that extends hit locus\n",
    "def extend_hit(chrNum, start, end, strand, extendLength = 1000):\n",
    "    chr_length = chromosome_sizes[chrNum]                              \n",
    "    move_start = start - extendLength\n",
    "    move_end = end + extendLength\n",
    "    annotation = \"both\"\n",
    "\n",
    "    if move_start < 1 and move_end > chr_length:\n",
    "        return [1, chr_length, \"last extension\"]\n",
    "    else:\n",
    "        if move_start < 1: \n",
    "            move_start = 1\n",
    "            annotation = \"upstream\"\n",
    "        elif move_end > chr_length: \n",
    "            move_end = chr_length\n",
    "            annotation = \"downstream\"\n",
    "            \n",
    "    return [move_start, move_end, annotation]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2652d614-d05b-47aa-ba13-34bfd905eaa7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene의 TSS, and distance\n",
    "def check_distance(gene_id, hit_locus):\n",
    "    \n",
    "    # create locus from gene_id\n",
    "    gene = genome.gene_by_id(gene_id)\n",
    "    gene_locus = pyensembl.Locus(contig=gene.contig, start=gene.start, end=gene.end, strand=gene.strand)\n",
    "    if gene.strand == \"+\":\n",
    "        TSS_point = gene.start\n",
    "    else:    \n",
    "        TSS_point = gene.end\n",
    "\n",
    "    # calculate the distance\n",
    "    # distance = hit_locus.distance_to_locus(gene_locus)              # if strand is different, (+/-) they calculate distance as infinity.\n",
    "    if hit_locus.distance_to_interval(gene.start, gene.end) == 0:\n",
    "        distance = 0\n",
    "    else: \n",
    "        distance = hit_locus.distance_to_interval(TSS_point, TSS_point)\n",
    "    \n",
    "    return distance"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "977bedad-9bec-4bd0-aad4-9ce69e15b7df",
   "metadata": {},
   "source": [
    "## Main1 - using Carsten's' Gene annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2daa4646-6ee1-4975-b629-06372aec60e7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "41"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "similar_pcg_set = set()\n",
    "\n",
    "i = 0\n",
    "while i < len(blast_Actg1): \n",
    "    row = blast_Actg1.iloc[i]\n",
    "\n",
    "    hit_chrNum = row['subject id'][3:]\n",
    "    hit_start = int(row['s. start'])\n",
    "    hit_end =  int(row['s. end'])\n",
    "    hit_strand = row['subject strand']\n",
    "\n",
    "    # create hit Locus object\n",
    "    hit_locus = pyensembl.Locus(contig=hit_chrNum, start=hit_start, end=hit_end, strand=hit_strand)\n",
    "\n",
    "    gene_id = []\n",
    "    gene_type = []\n",
    "    if pd.notna(row['gene_id']):\n",
    "        gene_id = row['gene_id'].split(',')\n",
    "    if pd.notna(row['gene_type']):\n",
    "        gene_type = row['gene_type'].split(',')\n",
    "    \n",
    "    for j in range(0, len(gene_id)):\n",
    "#         if check_distance(gene_id[j], hit_locus) < 2000 and gene_type[j] == 'protein_coding':   # same result\n",
    "        if gene_type[j] == 'protein_coding':                                                      # same result\n",
    "            similar_pcg_set.add(gene_id[j])\n",
    "\n",
    "    i += 1\n",
    "\n",
    "len(similar_pcg_set)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "485ea99d-0f66-487c-8c96-adb86ef9326b",
   "metadata": {},
   "source": [
    "## Main2 - add 2kb distance at hit locus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "014f370d-e777-4c20-9a89-bcae7a48cbbd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hit info:  7 5128966 5129315 - \n",
      "gene info:  5129291 5149670 - ENSMUSG00000118215 0\n",
      "hit info:  7 5128966 5129315 - \n",
      "gene info:  5129291 5149670 - ENSMUSG00000118215 0\n",
      "hit info:  8 123893148 123893467 - \n",
      "gene info:  119910841 124345722 + ENSMUSG00000092329 0\n",
      "hit info:  8 123893148 123893467 - \n",
      "gene info:  119910841 124345722 + ENSMUSG00000092329 0\n",
      "hit info:  8 123893148 123893467 - \n",
      "gene info:  119910841 124345722 + ENSMUSG00000092329 0\n",
      "hit info:  8 123892009 123892152 - \n",
      "gene info:  119910841 124345722 + ENSMUSG00000092329 0\n",
      "hit info:  8 123892009 123892152 - \n",
      "gene info:  119910841 124345722 + ENSMUSG00000092329 0\n",
      "hit info:  8 123892009 123892152 - \n",
      "gene info:  119910841 124345722 + ENSMUSG00000092329 0\n",
      "hit info:  8 123892549 123892724 - \n",
      "gene info:  119910841 124345722 + ENSMUSG00000092329 0\n",
      "hit info:  8 123892549 123892724 - \n",
      "gene info:  119910841 124345722 + ENSMUSG00000092329 0\n",
      "hit info:  8 123892549 123892724 - \n",
      "gene info:  119910841 124345722 + ENSMUSG00000092329 0\n",
      "hit info:  8 123892850 123892999 - \n",
      "gene info:  119910841 124345722 + ENSMUSG00000092329 0\n",
      "hit info:  8 123892850 123892999 - \n",
      "gene info:  119910841 124345722 + ENSMUSG00000092329 0\n",
      "hit info:  8 123892850 123892999 - \n",
      "gene info:  119910841 124345722 + ENSMUSG00000092329 0\n",
      "hit info:  8 123892233 123892408 - \n",
      "gene info:  119910841 124345722 + ENSMUSG00000092329 0\n",
      "hit info:  8 123892233 123892408 - \n",
      "gene info:  119910841 124345722 + ENSMUSG00000092329 0\n",
      "hit info:  8 123892233 123892408 - \n",
      "gene info:  119910841 124345722 + ENSMUSG00000092329 0\n",
      "hit info:  8 123893568 123893662 - \n",
      "gene info:  119910841 124345722 + ENSMUSG00000092329 0\n",
      "hit info:  8 123893568 123893662 - \n",
      "gene info:  119910841 124345722 + ENSMUSG00000092329 0\n",
      "hit info:  8 123893568 123893662 - \n",
      "gene info:  119910841 124345722 + ENSMUSG00000092329 0\n",
      "hit info:  16 67238129 67238266 - \n",
      "gene info:  66655421 67620908 - ENSMUSG00000064115 0\n",
      "hit info:  16 67238129 67238266 - \n",
      "gene info:  66655421 67620908 - ENSMUSG00000064115 0\n",
      "hit info:  16 67238129 67238266 - \n",
      "gene info:  66655421 67620908 - ENSMUSG00000064115 0\n",
      "hit info:  7 5129368 5129419 - \n",
      "gene info:  5129291 5149670 - ENSMUSG00000118215 0\n",
      "hit info:  7 5129368 5129419 - \n",
      "gene info:  5129291 5149670 - ENSMUSG00000118215 0\n",
      "hit info:  9 97596930 97597024 + \n",
      "gene info:  97444395 98033181 - ENSMUSG00000032452 0\n",
      "hit info:  9 97596930 97597024 + \n",
      "gene info:  97444395 98033181 - ENSMUSG00000032452 0\n",
      "hit info:  9 97596930 97597024 + \n",
      "gene info:  97444395 98033181 - ENSMUSG00000032452 0\n",
      "hit info:  7 5129317 5129357 - \n",
      "gene info:  5129291 5149670 - ENSMUSG00000118215 0\n",
      "hit info:  7 5129317 5129357 - \n",
      "gene info:  5129291 5149670 - ENSMUSG00000118215 0\n",
      "45\n"
     ]
    }
   ],
   "source": [
    "similar_pcg_set = set()\n",
    "\n",
    "i = 0\n",
    "while i < len(blast_Actg1): \n",
    "    row = blast_Actg1.iloc[i]\n",
    "\n",
    "    hit_chrNum = row['subject id'][3:]\n",
    "    hit_start = int(row['s. start'])\n",
    "    hit_end =  int(row['s. end'])\n",
    "    hit_strand = row['subject strand']\n",
    "\n",
    "    # create hit Locus object\n",
    "    hit_locus = pyensembl.Locus(contig=hit_chrNum, start=hit_start, end=hit_end, strand=hit_strand)\n",
    "\n",
    "    extend = \"NA\"\n",
    "    while (extend != \"false\"):\n",
    "        extend_hit_result = extend_hit(hit_chrNum, hit_start, hit_end, hit_strand, 2000)    # default extension\n",
    "        hit_start = extend_hit_result[0]\n",
    "        hit_end = extend_hit_result[1]\n",
    "        gene_in_locus_list = find_gene_in_locus(hit_chrNum, hit_start, hit_end, hit_strand)             # find gene in locus\n",
    "        if len(gene_in_locus_list) == 0:                                                                # if there is no gene in locus, extend hit length\n",
    "            extend_hit_result = extend_hit(hit_chrNum, hit_start, hit_end, hit_strand, 1000)\n",
    "            hit_start = extend_hit_result[0]\n",
    "            hit_end = extend_hit_result[1]\n",
    "            extend = extend_hit_result[2]\n",
    "        else:                                                                                            # if there is gene in locus, print the gene id and distance\n",
    "            # check distance between hit locus and gene\n",
    "            info = \"\"\n",
    "            for gene_id in gene_in_locus_list: \n",
    "                distance = check_distance(gene_id, hit_locus)\n",
    "                gene = genome.gene_by_id(gene_id);\n",
    "                info += f\"{gene_id} - Distance: {distance}; \"\n",
    "                if distance < 2000 and gene.biotype == \"protein_coding\":\n",
    "                    #################\n",
    "                    if gene_id == \"ENSMUSG00000032452\" or gene_id == \"ENSMUSG00000064115\" or gene_id == \"ENSMUSG00000092329\" or gene_id == \"ENSMUSG00000118215\":\n",
    "                        print(\"hit info: \", hit_locus.contig, hit_locus.start, hit_locus.end, hit_locus.strand,\n",
    "                              \"\\ngene info: \", gene.start, gene.end, gene.strand, gene_id, distance)\n",
    "                    #################\n",
    "                    similar_pcg_set.add(gene_id)\n",
    "            \n",
    "            # Update DataFrame, add gene and extend information\n",
    "            # blast_Actg1_local_ws4.at[i, 'genes'] = info\n",
    "            # blast_Actg1_local_ws4.at[i, 'extend'] = extend\n",
    "            extend = 'false'\n",
    "\n",
    "    i += 1\n",
    "\n",
    "print(len(similar_pcg_set))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8be78fcf-86e4-4506-a285-3738537efa5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"/home/ylee/blast/result/\")\n",
    "\n",
    "pcg_ids = [gene.gene_id for gene in genome.genes() if gene.biotype == 'protein_coding']\n",
    "pcg_set = set(pcg_ids)\n",
    "nonsimilar_pcg_set = pcg_set - similar_pcg_set\n",
    "\n",
    "# Save the set as a pickle file\n",
    "with open('similar_pcg_set_C.pkl', 'wb') as file:\n",
    "    pickle.dump(similar_pcg_set, file)\n",
    "\n",
    "with open('nonsimilar_pcg_set_C.pkl', 'wb') as file:\n",
    "    pickle.dump(nonsimilar_pcg_set, file)\n",
    "\n",
    "with open('pcg_set.pkl', 'wb') as file:\n",
    "    pickle.dump(pcg_set, file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29fa400e-4c6a-4251-a564-40dbec532cbb",
   "metadata": {},
   "source": [
    "## Check difference between result of J and K"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "4e87edfb-eed3-4c9d-8084-2851df37d360",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('similar_pcg_set_Y.pkl', 'rb') as file:\n",
    "    similar_pcg_set_Y = pickle.load(file)\n",
    "\n",
    "with open('nonsimilar_pcg_set_Y.pkl', 'rb') as file:\n",
    "    nonsimilar_pcg_set_Y = pickle.load(file)\n",
    "\n",
    "with open('similar_pcg_set_C.pkl', 'rb') as file:\n",
    "    similar_pcg_set_C = pickle.load(file)\n",
    "\n",
    "with open('nonsimilar_pcg_set_C.pkl', 'rb') as file:\n",
    "    nonsimilar_pcg_set_C = pickle.load(file)\n",
    "\n",
    "with open('pcg_set.pkl', 'rb') as file:             \n",
    "    pcg_set = pickle.load(file)                        # from pyensembl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "472f2614-18ca-42cc-b8dc-31562f66deef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ENSMUSG00000008658',\n",
       " 'ENSMUSG00000025812',\n",
       " 'ENSMUSG00000032452',\n",
       " 'ENSMUSG00000062825',\n",
       " 'ENSMUSG00000064115',\n",
       " 'ENSMUSG00000070979',\n",
       " 'ENSMUSG00000092329',\n",
       " 'ENSMUSG00000118215'}"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "similar_pcg_set_Y - similar_pcg_set_C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "5e0dcf55-3b27-4b83-9df4-994d0d7d70ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ENSMUSG00000023826'}"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "similar_pcg_set_C - similar_pcg_set_Y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "e886ded5-8092-4774-a5a7-cf93f1373097",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ENSMUSG00000032452',\n",
       " 'ENSMUSG00000064115',\n",
       " 'ENSMUSG00000092329',\n",
       " 'ENSMUSG00000118215'}"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "similar_pcg_set - similar_pcg_set_C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "d933b9e8-fef7-41ba-9c90-bf9cf6cc1b0a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ENSMUSG00000008658',\n",
       " 'ENSMUSG00000025812',\n",
       " 'ENSMUSG00000062825',\n",
       " 'ENSMUSG00000070979'}"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "similar_pcg_set_Y - similar_pcg_set"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd7c4c00-10bc-4841-ad35-1958eb478a7f",
   "metadata": {},
   "source": [
    "## Codes from K *need modification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "4a5a17fc-54cf-414f-a008-12d3e967b06d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "41\n"
     ]
    }
   ],
   "source": [
    "Carsten_annotated = pd.read_csv(\"ENSMUST00000071555.12.transcript.blastn.tsv.parsed.anno.final\", sep=\"\\t\", header=0)\n",
    "\n",
    "def check_overlapped(start_1, end_1, start_2, end_2):\n",
    "    if end_2 >= start_1 or end_1 >= start_2:\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "final_ls = []\n",
    "all_ls = [] \n",
    "other_ls = []\n",
    "for i in range(len(Carsten_annotated)):\n",
    "    chr_ = Carsten_annotated.iloc[i]['subject id']\n",
    "    start = Carsten_annotated.iloc[i]['s. start'] \n",
    "    end = Carsten_annotated.iloc[i]['s. end']\n",
    "    item = Carsten_annotated.iloc[i]['gene_id']\n",
    "    if isinstance(item, str):\n",
    "        ls = item.split(',')\n",
    "        \n",
    "        for gene_id in ls:\n",
    "            gene = genome.gene_by_id(gene_id)\n",
    "            g_start = gene.start  \n",
    "            g_end = gene.end\n",
    "            if check_overlapped(g_start, g_end, start, end): # if overlapped with gene body\n",
    "                final_ls.append(gene_id)\n",
    "            else: #check if overlapped with TSS to -2000bp upstream region\n",
    "                if gene.strand == '-':\n",
    "                    TSS = gene.end\n",
    "                    if end - TSS <=2000 or start - TSS <=2000:\n",
    "                        final_ls.append(gene_id)\n",
    "                    else:\n",
    "                        other_ls.append((gene_id, TSS, chr_, start, end))  \n",
    "                    \n",
    "                else:\n",
    "                    TSS = gene.start\n",
    "                    if TSS - end <=2000 or TSS - start<=2000:\n",
    "                        final_ls.append(gene_id)\n",
    "                    else:\n",
    "                        other_ls.append((gene_id, TSS, chr_, start, end))  \n",
    "                \n",
    "          \n",
    "            all_ls.append(gene_id)\n",
    "            \n",
    "protein_coding_ls = []\n",
    "for gene_id in list(set(final_ls)): \n",
    "    if genome.gene_by_id(gene_id).biotype =='protein_coding':\n",
    "        protein_coding_ls.append(gene_id)\n",
    "\n",
    "similar_pcg_set_K = set(protein_coding_ls)\n",
    "print(len(similar_pcg_set_K))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py_3.8.13",
   "language": "python",
   "name": "py_3.8.13."
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
